### DISCUSSION TITLE
Teensy as a Control Surface in Logic Pro X: The Epic Conundrum

### DISCUSSION THREAD

**Discussion by hali1850** [tttapa/Control-Surface#734] [Q&A] [answered]
Hi,

I have successfully created a simple MIDI controller consisting of 2 potentiometers (both rotary for now, but intended final setup is to have one rotary and one slide potentiometer) and 3 LED's corresponding to 3 transport buttons (play, record, toggle cycle mode) using my Teensy LC. I have a total of 8 buttons, each for different purposes.

Now, my problem is a simple problem: I want to be able to plug my Teensy control surface/MIDI controller into my computer, open up my DAW, not have to fuss too much with controls and it be automatically mapped to control Logic in the following ways -
 
1) Transport controls are synced up meaning when I press play button on my Teensy controller, the associated LED lights up and Logic plays. However, if I also press play in the DAW, the LED on my controller also lights up. Conversely, when I stop playback in Logic, the LED turns off (or more desirably, dims.) This feature should ideally extend to my Mute, Solo, Record, Cycle, Read/Latch Automation toggle, next channel and previous channel functions. 
2) My rotary potentiometer controls the currently selected channel's pan
3) My slide potentiometer controls the currently selected channel's volume fader. 

From this description, I hope you can easily deduce the type of device I'm trying to build for my home studio. The problem seems simple, but it ofc is not.

From my research and understanding so far, this plug-and-play, control surface functionality (typically) requires the use of the Mackie Control protocol/algorithm. 
I understand that this in essence is a proprietary mapping algorithm of the standard MIDI protocol which helps a DAW to communicate with a piece of hardware like the Mackie Control, and also conventional commercial MIDI controller products (although I have read that those guys use proprietary software and protocol to get these things done, on which they collude with the DAW-makers to make it 'plug-and-play' as their controller appear as a standard selectable controller in the list and are subsequently easy to setup and so on.)

I need my little Teensy to be able to masquerade as a Mackie Control, making my DAW automatically recognise its functionality and therefore me not having to manually map my parameters with simple CC after initially coding it; but also crucially allowing for there to be communication between my Teensy and Logic in order to update and sync with the LED's aboard. I understand that on the Getting Started page of this great Control-Surface repository is a reassurance that the library integrates a large part of the Mackie Control Universal protocol, but I'm still confused as to how to accomplish my desired feature list above utilising this code+protocol, and setting up a two-way communication between my DAW and my Teensy in a way which does not require external software like LoopMIDI and general clutter. Just the way my actual MIDI controllers are plug and play and do not require external fuss (Native Instruments A61, Arturia Keystep but not so much), my Teensy project should be too. I am hopeful this is possible.

I am aware that in order for the DAW to update my LED's when it is doing something related to each of them, that it needs to exchange data with the Teensy of course.
I did some diving using 'MIDI Monitor' and 'MidiView' software in order to intercept the messages that Logic Pro sends when I engage the various functions of the DAW which I desire to capture and utilise on my controller. I first setup a Mackie Control as a control surface in the 'Control Surface Setup' as I think this is a good starting place. I then set the output to MidiView in order to intercept and read the messages using the software. Although I can't be sure if the messages I intercepted are helpful in my endeavour, I'll tread this path for now.

Here are my findings:

Play – Note On, A#5, Velocity 127

Stop – Note Off, A#5, Velocity 0 

Record – Note On, F5, Velocity 127
            + Note On, B5, Velocity 127

Solo ON – Note On, G7, Velocity 1
              + Note On, G#-2 Velocity 127

Solo OFF –  Note Off, G7, Velocity 0
               +  Note off G#-2 Velocity 0

And so on and so forth.

I extracted this information as carefully as I could, as when pressing any of these functions, Logic spits out a flurry mess of messages, all of which I gather are pertaining to the position of the playhead as I'm able to replicate the influx of these frenzied messages by moving the playhead around. See attached the .csv file log from MiDiView of one such example. [Play then Stop.csv] (https://github.com/tttapa/Control-Surface/files/8502640/Play.then.Stop.csv)

(EDIT: By monitoring the MIDI out within Logic Pro itself I was able to further refine and double down on this list)

By using this website: https://www.inspiredacoustics.com/en/MIDI_note_numbers_and_center_frequencies I was able to cross reference the MIDI values of each of these notes. Curiously, when I set MidiView to show me the message numbers instead of notes, one of these two sources are one octave off. For example, MidiView says that G7, which is Solo ON, is 115. However, the chart says 115 is G8.
https://sites.google.com/view/mackiecontroluniversaldiyguide/home This is the website I used to then reference what the protocol itself mandates for these controls. This documentation says that Play is A#6, whereas I have observed Logic sending out one octave lower at A#5 for Play. I have to read somewhere to expect this octave shift in the MIDI protocol or something to that effect, but haven't had time yet to compute what it means and how this will affect my endeavour. 

And so, armed with this information, I reckon I now have what I assume are the essential bits of data in order to begin trying to sync Logic to my Teensy control surface/MIDI controller. 

I would appreciate some guidance on how or if I can achieve this functionality using the Control Surface library, such that my feature list is complete and my MIDI runs smoothly and controls my DAW + received feedback from it. 

Sidenote: This video proved to be quite illuminating, down this otherwise dimly lit path. https://www.youtube.com/watch?v=36Ek5qyTmTk&list=PL78elaZFXOd65nN4ZGvJZlOi3cRB-Bd8B

Here is my yet simple code, which only makes my Teensy a conventional and simple MIDI controller so far:

`#include <Control_Surface.h> // Include the Control Surface library
 
// Instantiate a MIDI over USB interface.
USBMIDI_Interface midi;
 
// Instantiate a CCButton object
CCButtons<10> buttons = {
  {3, 6, 9, 2, 5, 10, 11, 7, 0, 1}, // button pins
  {MIDI_CC::General_Purpose_Controller_1, CHANNEL_1}, // address of first button
  {1, 0}, // interval between addresses (1 controller number, 0 channels)
};

// Instantiate an array of CCPotentiometer objects
CCPotentiometer potentiometers[] {
  {A8 ,        // Analog pin connected to potentiometer 1
   0x30},     // Controller number of the first potentiometer
  {A7,        // Analog pin connected to potentiometer 1
   0x29},     // Controller number of the first potentiometer
};

#include <Arduino_Helpers.h>
 
#include <AH/Hardware/Button.hpp>
 
// Create a Button object that reads a push button connected to pin 2:
Button pushbutton1 {3};
Button pushbutton2 {6};
Button pushbutton3 {9};
 
// The pin with the LED connected:
const pin_t ledPin1 = {20};
const pin_t ledPin2 = {19};
const pin_t ledPin3 = {18};

void setup() {
  Control_Surface.begin(); // Initialize Control Surface
  
  pinMode(ledPin1, OUTPUT);
  pushbutton1.begin();
  pinMode(ledPin2, OUTPUT);
  pushbutton2.begin();
  pinMode(ledPin3, OUTPUT);
  pushbutton3.begin();
  // You can invert the input, for use with normally closed (NC) switches:
  // pushbutton.invert();
};

void loop() {
  static bool ledState1 = HIGH;
  // Read the digital input, debounce the signal, and check the state of
  // the button:
  if (pushbutton1.update() == Button::Falling) {
    ledState1 = !ledState1; // Invert the state of the LED
    // Update the LED with the new state
    digitalWrite(ledPin1, ledState1 ? LOW : HIGH);
  }
  // Read the digital input, debounce the signal, and check the state of
  // the button:
    static bool ledState2 = HIGH;
  if (pushbutton2.update() == Button::Falling) {
    ledState2 = !ledState2; // Invert the state of the LED
    // Update the LED with the new state
    digitalWrite(ledPin2, ledState2 ? LOW : HIGH);
  }
  // Read the digital input, debounce the signal, and check the state of
  // the button:
     static bool ledState3 = HIGH; 
  if (pushbutton3.update() == Button::Falling) {
    ledState3 = !ledState3; // Invert the state of the LED
    // Update the LED with the new state
    digitalWrite(ledPin3, ledState3 ? LOW : HIGH);        
  }
  Control_Surface.loop(); // Update the Control Surface
}   `

I would appreciate any and all help guidance and expertise to help me get my project across the line. 
Warm regards
H

**Comment by Veroledez** [comment, marked as answer]
Hola, el protocolo mackie  ya esta soportado por esta gran libreria, asi  solo debes usar en los constructores lo que necesitas, por ejemplo en lugar de MIDI::General_purpose, usa MCU::PLAY, MCU::STOP, MCU:: SOLO_1,  busca en la libreria un archivo MCU ahi esta todas las declaraciones que buscas aplica para leds tambien, ya solo tendrias  que en el daw selecccionar el script mackie control
